lib/ruby_smb/dcerpc/lsarpc.rb

```diff
@@ -1,11 +1,644 @@
+          raise RubySMB::Dcerpc::Error::LsarpcError,
+                "Error returned while opening policy: "\
+                  "#{WindowsError::NTStatus.find_by_retval(lsar_response.error_status.value).join(',')}"
+        end
+        lsar_response.policy_handle
+      end
+
+      def lsar_query_information_policy(policy_handle:, information_class:)
+        lsar_request = LsarQueryInformationPolicyRequest.new(
+          policy_handle: policy_handle,
+          information_class: information_class
+        )
+        response = dcerpc_request(lsar_request)
+        begin
+          lsar_response = LsarQueryInformationPolicyResponse.read(response)
+        rescue IOError
+          raise RubySMB::Dcerpc::Error::InvalidPacket, 'Error reading LsarQueryInformationPolicyResponse'
+        end
+        unless lsar_response.error_status == WindowsError::NTStatus::STATUS_SUCCESS
+          raise RubySMB::Dcerpc::Error::LsarpcError,
+                "Error returned while querying domain information: "\
+                  "#{WindowsError::NTStatus.find_by_retval(lsar_response.error_status.value).join(',')}"
+        end
+        lsar_response.policy_information
+      end
+
+      def lsar_query_information_policy_2(policy_handle:, information_class:)
+        lsar_request = LsarQueryInformationPolicy2Request.new(
+          policy_handle: policy_handle,
+          information_class: information_class
+        )
+        response = dcerpc_request(lsar_request)
+        begin
+          lsar_response = LsarQueryInformationPolicy2Response.read(response)
+        rescue IOError
+          raise RubySMB::Dcerpc::Error::InvalidPacket, 'Error reading LsarQueryInformationPolicy2Response'
+        end
+        unless lsar_response.error_status == WindowsError::NTStatus::STATUS_SUCCESS
+          raise RubySMB::Dcerpc::Error::LsarpcError,
+                "Error returned while querying domain information: "\
+                  "#{WindowsError::NTStatus.find_by_retval(lsar_response.error_status.value).join(',')}"
+        end
+        lsar_response.policy_information
+      end
+
+      def lsar_close_handle(policy_handle:)
+        lsar_request = LsarCloseHandleRequest.new(
+          policy_handle: policy_handle
+        )
+        response = dcerpc_request(lsar_request)
+        begin
+          lsar_response = LsarCloseHandleResponse.read(response)
+        rescue IOError
+          raise RubySMB::Dcerpc::Error::InvalidPacket, 'Error reading LsarCloseHandleResponse'
+        end
+        unless lsar_response.error_status == WindowsError::NTStatus::STATUS_SUCCESS
+          raise RubySMB::Dcerpc::Error::LsarpcError,
+                "Error returned while closing policy handle: "\
+                  "#{WindowsError::NTStatus.find_by_retval(lsar_response.error_status.value).join(',')}"
+        end
+        lsar_response.policy_handle
+      end
+
+      def lsar_lookup_sids(policy_handle:, sid_enum_buffer:, lookup_level:)
```

Comment:
I think I commented on this else where but it would be easier to sue if the caller didn't need to deal constructing the sid_enum_buffer themselves but instead could pass an array of SID strings and this function would translate that into the necessary object instance for BinData.